homeworkW12D03/src/ba/bitcamp/homeworkW12D03/task01/CountLettersFromFile.java

```diff
@@ -0,0 +1,139 @@
+package ba.bitcamp.homeworkW12D03.task01;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * 
+ * @author Gordan
+ *
+ *         This class represents one program that counts how many time first
+ *         letter from text file repeats in same text file.
+ */
+public class CountLettersFromFile {
+
+	// Defining variables
+	private static int counter = 1;
+	private static LinkedBlockingQueue<Master> assignments;
+	private static ArrayList<Worker> workers;
+	private static Object lock = new Object();
+
+	public static void main(String[] args) {
+
+		// Initializing list and queue
+		assignments = new LinkedBlockingQueue<>();
+		workers = new ArrayList<>();
+
+		try {
+
+			// Creating reader for reading text file
+			BufferedReader reader = new BufferedReader(new FileReader(new File("Text.txt")));
+			// Taking first letter from file
+			char firstChar = (char) reader.read();
+			// Putting letter in low case
+			firstChar = Character.toLowerCase(firstChar);
+			// Printing first letter
+			System.out.println(firstChar);
+
+			String line = "";
+
+			// Reading the rest of the text
+			while (reader.ready()) {
+				// Taking line and making it low case
+				line = reader.readLine().toLowerCase();
+				// Making master thread
+				Master master = new Master(line, firstChar);
+				// Adding master thread to queue of threads
+				assignments.add(master);
+			}
+
+			reader.close();
+
+			// Starting time
+			long start = System.currentTimeMillis();
+
+			// Creating four worker threads, starting and adding them to the
+			// list of
+			// workers
+			for (int i = 0; i < 4; i++) {
+				Worker worker = new Worker();
+				worker.start();
+				workers.add(worker);
+			}
+
+			// Making each of them wait for previous thread to finish
+			for (Worker c : workers) {
+				try {
+					c.join();
+				} catch (InterruptedException e) {
+					e.printStackTrace();
+				}
+			}
+
+			// Printing information of time necessary to finish and repeating
+			// letter
+			System.out.println("Time: " + (System.currentTimeMillis() - start) + " [ms]");
+			System.out.printf("Letter '%s' repeats %d times.", firstChar, counter);
+
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * 
+	 * @author Gordan
+	 * 
+	 *         This is inner class that represents a working thread. It gets
+	 *         assignment from master thread and finishes it.
+	 */
+	private static class Worker extends Thread {
+
+		public void run() {
```

Comment:
This thread can get stuck and will never finish, here is why:

- Imagine two threads running in parallel and only one assignment is left in the queue
- Both threads test if assignment container is empty and false is returned
- Both threads go to the `assignments.take()` line

Since `take()` is implemented properly, only one thread will obtain remaining task, the other thread will get stuck because `take()` is a blocking method. Blocking means that code execution in the thread that invoked the method will wait until new element is available or thread is interrupted. Since there is no interrupt anywhere, code will never finish executing.

Please see `poll` method; it offers a mechanism for specifying maximum time to wait for an element to be available, this will allow all the threads to finish.